azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +101,232 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+            individual_policy_schema = _POLICY_SCHEMA["extensionPolicies"]["extensions"]["<extensionName>"]
+            if not isinstance(individual_policy, dict):
+                raise InvalidPolicyError("invalid type {0} for attribute '{1}', please change to object."
+                                         .format(type(individual_policy).__name__, extension_name))
+            self.__validate_schema(individual_policy, individual_policy_schema)
+
+            extension_signature_policy = self.__parse_attribute(individual_policy, "signatureRequired", individual_policy_schema, extension_name)
+            if extension_signature_policy is None:
+                extension_signature_policy = self._policy["extensionPolicies"]["signatureRequired"]
+
+            policy_to_add = {
+                "signatureRequired": extension_signature_policy
+            }
+            parsed_extensions_dict[extension_name] = policy_to_add
+
+        # Convert "extensions" to a case-folded dict for case-insensitive lookup
+        case_folded_extensions_dict = _CaseFoldedDict.from_dict(parsed_extensions_dict)
+        self._policy["extensionPolicies"]["extensions"] = case_folded_extensions_dict
 
     @staticmethod
-    def is_policy_enforcement_enabled():
+    def __validate_schema(policy, schema):
         """
-        Check whether user has opted into policy enforcement feature.
-        Caller function should check this before performing any operations.
+        Validate the provided policy against the schema - we only do a shallow check (no recursion into nested dicts).
+        If there is an unrecognized attribute, raise an error.
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
-        # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        for key in policy:
+            if key not in schema:
+                raise InvalidPolicyError("attribute '{0}' is not defined in the policy schema. Please refer to the policy documentation "
+                                         "and change or remove this attribute accordingly.".format(key))
+
+    @staticmethod
+    def __parse_attribute(policy, key, schema, parent_attribute_name=None):
+        """
+        Return policy[key], return None if the key does not exist in the policy.
+
+        Raise error if policy[key] does not match the expected value specified in the schema (schema[key])
+        For attributes that are used in multiple places (ex: "signatureRequired"), specify the parent attribute in
+        the error message to avoid ambiguity.
+        """
+        value = policy.get(key)
+        if value is None:
+            return None
+
+        expected_type = schema.get(key)
+
+        # don't recurse into nested dicts, just validate that the type is dict
+        if isinstance(expected_type, dict):
+            expected_type = dict
```

Comment:
what is the intention of lines 266, 267?

Reply:
duplicate comment